### PR TITLE
fix: disregarding already booked spots or blocked calendar times

### LINKED ISSUE #1271: Disregarding already booked spots or blocked calendar times
### Issue Summary

Current events are disregarding already booked spots and blocked calendar times. This means bookers can double-book and generally don't know when a time is taken.

### Steps to Reproduce

1. Have a day fully blocked in your calendar
2. Head over to booking (https://cal.com/dom/15min?date=2021-12-15%2B0100)
3. Time is available

Screenshot 2021-12-08 at 08 47 17
Screenshot 2021-12-08 at 08 47 05

### Technical details

- Browser version: Chrome 96.
- Node.js version: Using cal.com hosted
- GCal is connected in integrations

### PR DESCRIPTION
**Pull request by LouisHaftmann**
## What does this PR do?

Fixes double-booking issue

Fixes #1271, #1450

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How should this be tested?

1. Try booking the same time spot multiple times
2. Should displays error message

### CODE REVIEW

**Review by pumfleet — APPROVED** [review]

**Review by alannnc — APPROVED** [review]
Awesome! It's working properly now

**Review by agustif — COMMENTED** [review]
I think it's not working for me?
image

Maybe I'm doing something wrong. I just go back to the previous tab and book again and it works, no error.

I only see a change in an api endpoint, in the files changed, but I cannot reproduce/test it properly

**Comment by LouisHaftmann** [comment]
are you using any calendar integrations (google/apple/microsoft)?

**Comment by LouisHaftmann** [comment]
yea that's the problem, i'll look into it

**Comment by LouisHaftmann** [comment]
I just noticed that we aren't checking for conflicts before creating the booking, not even whether the time slot is out of bounds. Any reason for that?

**Comment by LouisHaftmann** [comment]
i believe [this line](https://github.com/calcom/cal.com/blob/8c4eed2bbc80f53a2fce96a490bf8cbd9f168ce9/apps/web/pages/api/book/event.ts#L309) is supposed create a unique seed for each time slot but why does it use `new Date().getTime()`?
```js
const seed = `${users[0].username}:${dayjs(req.body.start).utc().format()}:${new Date().getTime()}`;
```

**Comment by PeerRich** [comment]
since this touches quite a bit of core, @emrysal and @zomars can you test this?

**Comment by emrysal** [comment]
@LouisHaftmann 

> i believe [this line](https://github.com/calcom/cal.com/blob/8c4eed2bbc80f53a2fce96a490bf8cbd9f168ce9/apps/web/pages/api/book/event.ts#L309) is supposed create a unique seed for each time slot but why does it use `new Date().getTime()`?
> 
> ```js
> const seed = `${users[0].username}:${dayjs(req.body.start).utc().format()}:${new Date().getTime()}`;
> ```

You are entirely right about this; this is a bug that still exists and was a "hack" whilst integrating Daily to get it to work. (Not quite sure what broke, maybe @baileypumfleet remembers, or @lunchpaillola.

The idea of this seed before then was to prevent conflicts on database level (which was disabled causing the bugs). When this seed works, it would create a unique seed for each slot, causing a P2002 on database level (unique constraint violation). Database was originally deemed to be the ideal place of doing this as any other location would cause race conditions in times of high traffic.

**Comment by LouisHaftmann** [comment]
Even with the old seed it wouldn't be 100% safe because it doesn't include any info about end time or time slot duration. Changing an event type's duration (eg: from 30min to 15min) would break this conflict check because a booking could be created without conflict error although it conflicts with the old booking.

Example (event type duration: 30min):
1. Create booking at `1pm`
2. Trying to book event at `1pm` again will result in conflict
3. Change event type duration to `15min`
4. Trying to book event at `1pm` still results in conflict
5. Trying to book event at `1:15pm` works although it should conflict with old events

**Comment by LouisHaftmann** [comment]
In [this commit](https://github.com/calcom/cal.com/pull/2029/commits/73bc6dac17bf025068b9fc5cca7954e4aa8073d3) I made it so it fetches all existing bookings for the user and event type and checks for conflicts (including integration and video busy times) before even trying to create a booking in the db.

**Comment by emrysal** [comment]
Yeah this is an improvement on the current broken logic, but does need further development.

The event type duration isn't as much of a problem as the seed logic is intended to detect double bookings made at the same time, it's not meant to check freebusy - that happens before then.

Reasons for future (Approving this in principle) development are:
* Booking becomes slower because of conflict checking
* Higher chance of integration failure
* Still can be a race condition - but there is a smaller window.

Virtual calendar is a feature that will focus on improving this

**Review by emrysal — APPROVED** [review]

**Comment by LouisHaftmann** [comment]
A seed like
```js
userId:eventTypeId:startTime
```
would probably work for now. It doesn't cover all edge cases but would atleast prevent simple double booking

**Comment by LouisHaftmann** [comment]
Prisma's [@@unique](https://www.prisma.io/docs/reference/api-reference/prisma-schema-reference#unique-1) constraint could also work

**Comment by PeerRich** [comment]
> A seed like
> ```js
> userId:eventTypeId:startTime
> ```
> would probably work for now. It doesn't cover all edge cases but would atleast prevent simple double booking

once @zomars reviewed this PR, he can merge it

**Comment by zomars** [comment]
I'm trying to test locally and getting this issue when running `yarn test-e2e`:

```

@calcom/prisma:db-reset: $ yarn prisma db seed
@calcom/prisma:db-reset: $ /Users/zomars/Sites/cal.com/node_modules/.bin/prisma db seed
@calcom/prisma:db-reset: Environment variables loaded from .env
@calcom/prisma:db-reset: Running seed command `ts-node ./seed.ts` ...
@calcom/prisma:db-reset: PrismaClientKnownRequestError:
@calcom/prisma:db-reset: Invalid `prisma.user.upsert()` invocation:
@calcom/prisma:db-reset:
@calcom/prisma:db-reset:
@calcom/prisma:db-reset:   Unique constraint failed on the fields: (`username`)
@calcom/prisma:db-reset:     at cb (/Users/zomars/Sites/cal.com/node_modules/@prisma/client/runtime/index.js:38703:17)
@calcom/prisma:db-reset:     at async PrismaClient._request (/Users/zomars/Sites/cal.com/node_modules/@prisma/client/runtime/index.js:40859:18)
@calcom/prisma:db-reset:     at async createUserAndEventType (/Users/zomars/Sites/cal.com/packages/prisma/seed.ts:38:16)
@calcom/prisma:db-reset:     at async main (/Users/zomars/Sites/cal.com/packages/prisma/seed.ts:157:3) {
@calcom/prisma:db-reset:   code: 'P2002',
@calcom/prisma:db-reset:   clientVersion: '3.9.2',
@calcom/prisma:db-reset:   meta: { target: [ 'username' ] }
@calcom/prisma:db-reset: }
```

File: apps/web/pages/api/book/event.ts [hunk 1]
```diff
@@ -443,70 +428,96 @@ export default async function handler(req: NextApiRequest, res: NextApiResponse)
     });
 
     const credentials = currentUser.credentials;
+    const calendarBusyTimes: EventBusyDate[] = await prisma.booking
```
Comment (zomars) [inline apps/web/pages/api/book/event.ts:431]: Not a blocking issue but we should strive to rely on inferred types as much as possible.

File: apps/web/playwright/booking-pages.test.ts [hunk 2]
```diff
@@ -1,16 +1,80 @@
...
+    });
+
+    // return to same time spot booking page
+    await page.goto(bookingUrl);
+
+    // book same time spot again
+    await bookTimeSlot();
+
+    // check for error message
+    await expect(await page.locator("[data-testid=booking-fail]")).toBeVisible();
```
Comment (zomars) [inline apps/web/playwright/booking-pages.test.ts:75]: Uneeded await:

```suggestion
    await expect(page.locator("[data-testid=booking-fail]")).toBeVisible();
```

File: apps/web/playwright/booking-pages.test.ts [hunk 3]
```diff
@@ -31,11 +100,6 @@ test.describe("pro user", () => {
     await page.click('[data-testid="event-type-link"]');
     // Click [data-testid="incrementMonth"]
     await page.click('[data-testid="incrementMonth"]');
-
-    // @TODO: Find a better way to make test wait for full month change render to end
-    // so it can click up on the right day, also when resolve remove other todos
-    // Waiting for full month increment
-    await page.waitForTimeout(400);
```
Comment (zomars) [inline apps/web/playwright/booking-pages.test.ts:38]: I think this was removed by mistake. This is a temporal fix made by @alannnc so tests aren't randomly failing due to loading times.

**Review by zomars — CHANGES_REQUESTED** [review]
This is great! Just left some minor comments in there 🙏🏽

File: apps/web/playwright/booking-pages.test.ts (hunk 3 above)
Comment (LouisHaftmann) [reply apps/web/playwright/booking-pages.test.ts]: sry, removed this by accident

**Review by zomars — APPROVED** [review]
Pushed fixes to prevent further delays. Thank you for your contribution 🙏

**Comment by lunchpaillola** [comment]
> @LouisHaftmann
> 
> > i believe [this line](https://github.com/calcom/cal.com/blob/8c4eed2bbc80f53a2fce96a490bf8cbd9f168ce9/apps/web/pages/api/book/event.ts#L309) is supposed create a unique seed for each time slot but why does it use `new Date().getTime()`?
> > ```js
> > const seed = `${users[0].username}:${dayjs(req.body.start).utc().format()}:${new Date().getTime()}`;
> > ```
> 
> You are entirely right about this; this is a bug that still exists and was a "hack" whilst integrating Daily to get it to work. (Not quite sure what broke, maybe @baileypumfleet remembers, or @lunchpaillola.
> 
> The idea of this seed before then was to prevent conflicts on database level (which was disabled causing the bugs). When this seed works, it would create a unique seed for each slot, causing a P2002 on database level (unique constraint violation). Database was originally deemed to be the ideal place of doing this as any other location would cause race conditions in times of high traffic.
> 
> Edit: It used to be: https://github.com/calcom/cal.com/blame/58de9209512c45e253ba3737311a0e21e0d1cf10/pages/api/book/event.ts#L252

It's been a while so I don't know if that hack is still relevant since I'm sure the code has changed a bit. But for some historical context the new date had to do with rebooking a time that had been previously canceled. Without the hack if you had canceled a time you wouldn't be able to rebook that time even if the slot had been canceled.

Sounds like it's all sorted now though!

**Pull request merged**